common/unsafe/src/main/java/org/apache/spark/unsafe/types/VariantVal.java

```diff
@@ -0,0 +1,124 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.unsafe.types;
+
+import org.apache.spark.unsafe.Platform;
+
+import java.io.Serializable;
+import java.util.Arrays;
+
+/**
+ * The physical data representation of {@link org.apache.spark.sql.types.VariantType} that
+ * represents a semi-structured value. It consists of two binary values: {@link VariantVal#value}
+ * and {@link VariantVal#metadata}. The value encodes types and values, but not field names. The
+ * metadata currently contains a version flag and a list of field names. We can extend/modify the
+ * detailed binary format given the version flag.
+ * <p>
+ * A {@link VariantVal} can be produced by casting another value into the Variant type or parsing a
+ * JSON string in the {@link org.apache.spark.sql.catalyst.expressions.variant.ParseJson}
+ * expression. We can extract a path consisting of field names and array indices from it, cast it
+ * into a concrete data type, or rebuild a JSON string from it.
+ * <p>
+ * The storage layout of this class in {@link org.apache.spark.sql.catalyst.expressions.UnsafeRow}
+ * and {@link org.apache.spark.sql.catalyst.expressions.UnsafeArrayData} is: the fixed-size part is
+ * a long value "offsetAndSize". The upper 32 bits is the offset that points to the start position
+ * of the actual binary content. The lower 32 bits is the total length of the binary content. The
+ * binary content contains: 4 bytes representing the length of {@link VariantVal#value}, content of
+ * {@link VariantVal#value}, content of {@link VariantVal#metadata}. This is an internal and
+ * transient format and can be modified at any time.
+ */
+public class VariantVal implements Serializable {
+  protected final byte[] value;
+  protected final byte[] metadata;
+
+  public VariantVal(byte[] value, byte[] metadata) {
+    this.value = value;
+    this.metadata = metadata;
+  }
+
+  public byte[] getValue() {
+    return value;
+  }
+
+  public byte[] getMetadata() {
+    return metadata;
+  }
+
+  /**
+   * This function writes the binary content into {@code buffer} starting from {@code cursor}, as
+   * described in the class comment. The caller should guarantee there is enough space in `buffer`.
+   */
+  public void writeIntoUnsafeRow(byte[] buffer, long cursor) {
```

Comment:
```suggestion
  public void writeIntoUnsafeRow(Object baseObject, long baseOffset, long cursor) {
```

Reply:
there is no `byte[]` if Spark is using offheap mode.

Reply:
I'm not quite sure what you mean. This function is called by `UnsafeWriter`, which always uses `byte[]` to build an `UnsafeRow`. There is no benefit of changing this function to use a generic base object.

Reply:
interesting, `UnsafeWriter` always use `byte[]`, then it's fine.

Reply:
I'm a bit confused. Why not add `writeIntoUnsafeRow` into `UnsafeWriter`?

Reply:
Actually this makes sense to me. The job of writing to unsafe row belongs to `UnsafeWriter` and the code should be put there as well.

Reply:
It makes sense. The original reason I put `writeIntoUnsafeRow` in this class is to avoid code duplication (e.g., `readFromUnsafeRow` is called by two classes, such a duplication is quite common for other physical value types). But since `writeIntoUnsafeRow` only has one caller `UnsafeWriter`, it is okay to just put it in `UnsafeWriter`.